src/main/java/net/greghaines/jesque/worker/WorkerImpl.java

```diff
@@ -507,12 +507,14 @@ protected String getNextQueue() throws InterruptedException {
      */
     protected String pop(final String curQueue) {
         final String key = key(QUEUE, curQueue);
+        String inFlightKey = key(INFLIGHT, this.name, curQueue);
+
         switch (nextQueueStrategy) {
         case DRAIN_WHILE_MESSAGES_EXISTS:
-            return (String) this.jedis.evalsha(this.popScriptHash.get(), 3, key, key(INFLIGHT, this.name, curQueue),
+            return (String) this.jedis.evalsha(this.popScriptHash.get(), 3, key, inFlightKey,
                     JesqueUtils.createRecurringHashKey(key), Long.toString(System.currentTimeMillis()));
         case RESET_TO_HIGHEST_PRIORITY:
-            return (String) this.jedis.evalsha(this.multiPriorityQueuesScriptHash.get(), 1, curQueue);
+            return (String) this.jedis.evalsha(this.multiPriorityQueuesScriptHash.get(), 2, curQueue, inFlightKey, Long.toString(System.currentTimeMillis()));
```

Comment:
Both of you are right :-) However I am trying now to go with another option : instead of providing time as script parameter I am going to use redis internal **_time_** command. This way we will not depend on local time of each server in cluster running **_pop_** script but rather take time from redis server.